cmd/controller/app/controllers.go

```diff
@@ -64,28 +68,55 @@ func addControllers(mgr manager.Manager, client k8s.Client, informerFactory info
 			informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
 			informerFactory.KubeSphereSharedInformerFactory().Devops().V1alpha3().DevOpsProjects())
 
-		devopsPipelineController = pipeline.NewController(client.Kubernetes(),
-			client.KubeSphere(), devopsClient,
-			informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
-			informerFactory.KubeSphereSharedInformerFactory().Devops().V1alpha3().Pipelines())
-
 		devopsCredentialController = devopscredential.NewController(client.Kubernetes(),
 			devopsClient,
 			informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
 			informerFactory.KubernetesSharedInformerFactory().Core().V1().Secrets())
 
-		jenkinsConfigController = jenkinsconfig.NewController(&jenkinsconfig.ControllerOptions{
-			LimitRangeClient:    client.Kubernetes().CoreV1(),
-			ResourceQuotaClient: client.Kubernetes().CoreV1(),
-			ConfigMapClient:     client.Kubernetes().CoreV1(),
-
-			ConfigMapInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().ConfigMaps(),
-			NamespaceInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
-			InformerFactory:   informerFactory,
-
-			ConfigOperator:  devopsClient,
-			ReloadCasCDelay: s.JenkinsOptions.ReloadCasCDelay,
-		}, s.JenkinsOptions)
+		// Choose controllers of CRDs (Pipeline and PipelineRun),
+		// by the field `PipelineBackend`in options.DevOpsControllerManagerOptions
+		if s.PipelineBackend == "jenkins" {
+			klog.Info("Jenkins was chosen to be the pipeline backend.")
+			devopsPipelineController = pipeline.NewController(client.Kubernetes(),
+				client.KubeSphere(), devopsClient,
+				informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
+				informerFactory.KubeSphereSharedInformerFactory().Devops().V1alpha3().Pipelines())
+
+			jenkinsConfigController = jenkinsconfig.NewController(&jenkinsconfig.ControllerOptions{
+				LimitRangeClient:    client.Kubernetes().CoreV1(),
+				ResourceQuotaClient: client.Kubernetes().CoreV1(),
+				ConfigMapClient:     client.Kubernetes().CoreV1(),
+
+				ConfigMapInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().ConfigMaps(),
+				NamespaceInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
+				InformerFactory:   informerFactory,
+
+				ConfigOperator:  devopsClient,
+				ReloadCasCDelay: s.JenkinsOptions.ReloadCasCDelay,
+			}, s.JenkinsOptions)
+		} else if s.PipelineBackend == "tekton" {
+			klog.Info("Tekton was chosen to be the pipeline backend.")
```

Comment:
How about just print the log output one time? Remove line 79 and line 98. Print the pipeline backend before line 78?

Reply:
> How about just print the log output one time? Remove line 79 and line 98. Print the pipeline backend before line 78?

Alright, it is better to print the log information out of the condition `if` block than to print it in the two condition blocks.